src/codegen/optimizer.lisp

```diff
@@ -421,8 +424,10 @@ requires direct constructor calls."
              (alexandria:when-let*
                  ((name (node-rator-name node))
                   (code (tc:lookup-code env name :no-error t))
+                  (fun-env-entry (tc:lookup-function env name :no-error t))
+                  (inline-p (when fun-env-entry (tc:function-env-entry-inline-p fun-env-entry)))
                   (_    (and (node-abstraction-p code)
-                             (funcall heuristic code)
+                             (or inline-p (funcall heuristic code))
```

Comment:
Would it make sense to rewrite this condition as the following?

```lisp
                  (fun-env-entry (tc:lookup-function env name :no-error t))
                  (_    (and (node-abstraction-p code)
                             (or (tc:function-env-entry-inline-p fun-env-entry)
                                 (funcall heuristic code))
                             (<= (length call-stack)
                                 max-depth)
```

Reply:
Yeah, I think that is a good way to rewrite it so that it will work with the `when-let*`.
The only thing I'm unsure of is if there ever can be a case when there is an entry in the `code-environment` but not in the `function-environment`. (I can't tell easily just from looking at the code, since the code entries and function entries are added at very different places in the code.) If so, then in order to allow heuristic inlining of functions with code env entries but not function env entries, it seems like a separate let statement might be necessary, like
```lisp
                  (_    (and (node-abstraction-p code)
                             (or (alexandria:when-let (fun-env-entry (tc:lookup-function env name :no-error t))
                                   (tc:function-env-entry-inline-p fun-env-entry))
                                 (funcall heuristic code))
                             (<= (length call-stack)
                                 max-depth)
```
or if not, then your suggestion should work exactly as written, or even without the `:no-error t`.

Reply:
If I remove `:no-error t` in my suggestion, I obtain the error "Internal coalton bug: Unknown function TREE:INCREASING-ORDER" while compiling `library/ord-map.lisp`. So, there indeed seem to be cases with code env entries but not function env entries.

However, I note that your suggestion works even without the `:no-error t`. My hypothesis is there are fun-env-entries only when code is an abstraction node.

Reply:
Committed in https://github.com/coalton-lang/coalton/pull/1290/commits/849af1f4816a31fa85e78eb7114ded8a61669d24